### DISCUSSION TITLE
How to limit the request body property to certain values?

### DISCUSSION THREAD

**Discussion by lukasz-madon** [schemathesis/schemathesis#1090] [Q&A]
I'd like to limit tests request to not include `{"status": "Archived"}`. Only "Draft" and "WaitingForApproval" (POST request)

What is the best way of doing it? Is it possible to define exact example "Draft" or do I need to filter it out?

```
VersionStatus:
      type: string
      enum:
        - Draft
        - WaitingForApproval
        - Published
        - Archived
```
test code

```python
import os

import schemathesis
from hypothesis import settings, HealthCheck

from conftest import get_root_url

schema = schemathesis.from_path(
    os.path.join(os.path.dirname(__file__), "api_spec.yaml"), get_root_url()
)


def pred(val):
    """filter out chars that makes tests flaky
    . is used for nesting with marshmallow
    NULL char is fine string in python, but postgres doesn't like it
    """
    x = str(val).rstrip()
    return not ("." in x or "\x00" in x or "%00" in x)


@schema.hooks.register("before_generate_body")
def schema_hook(context, strategy):
    return strategy.filter(pred)


@schema.parametrize()
@settings(max_examples=50, suppress_health_check=(HealthCheck.too_slow,))
def test_no_server_errors(case, flask_test_app, cookies):
    # `requests` will make an appropriate call under the hood
    response = case.call(cookies=cookies)
    # You could use built-in checks
    case.validate_response(response)
    # Or assert the response manually
    assert response.status_code < 500
```

**Comment by Stranger6667** [comment]
Hi @lukasz-madon 

I see a few ways of doing so with their own pros & cons.

1. Modify the schema via the `before_load_schema` hook

You could manually change your API schema in place before it is loaded to Schemathesis:

```python
@schemathesis.register
def before_load_schema(context, schema) -> None:
    schema["paths"]["/<operation>"]["<method>"]["parameters"][0]["schema"]["enum"] = ["Draft", "WaitingForApproval"]
```

The example above should be adjusted according to your API spec version & the exact `VersionStatus` location within the schema.

**Pros**:
- You won't have to deal with this enum anywhere further; it is globally changed
- data generation is slightly faster

**Cons**:
- modifying the schema might be a little fiddly as you need to properly search for this definition which may require updates after schema changes
- Having it on the global level maybe not desired (e.g., if you want to have different test functions generate different values)

2. Use `map` in `before_generate_body`

You can modify generated data via this hook:

```python
def set_status(val):
    val["status"] = "Draft"

@schema.hooks.register("before_generate_body")
def schema_hook(context, strategy):
    return strategy.filter(pred).map(set_status)
```

**Pros**:
- The change has a well-defined small scope. You only have to change the generated object
- Changes can be applied per globally / per-schema / per-test function

**Cons**:
- You need to filter out API operations that should not be affected by this change and apply the modification only when it is needed, which might be a bit fiddly
- It is slightly less efficient, as Hypothesis spends time to generate those values first

3. Change the value in the test function body

```python
@schema.parametrize()
@settings(max_examples=50, suppress_health_check=(HealthCheck.too_slow,))
def test_no_server_errors(case, flask_test_app, cookies):
    case.body["status"] = "Draft"
    ...
```

Pros & cons are similar to (2).

4. Use `@schema.given`

Here you can add some custom strategies without interacting with the schema at all.

```python
from hypothesis import strategies as st

@schema.parametrize()
@schema.given(data=st.data())
@settings(max_examples=50, suppress_health_check=(HealthCheck.too_slow,))
def test_no_server_errors(data, case, flask_test_app, cookies):
    case.body["status"] = st.sampled_from(["Draft", "WaitingForApproval"])
    ...
```

**Pros**:
- You can generate random values that will be integrated into the Hypothesis machinery
- You can use an arbitrary Hypothesis strategy, which gives a lot of flexibility

**Cons**:
- Still some overhead for data generation

Personally, (4) is my favorite :) Let me know if these options work for you.

**Reply by lukasz-madon** [reply]
Thanks for the reply. 

One questions. Option 2,3,4 (`case.body["status"] =`). Does it mean that we will set status for all cases? This single test is for the whole file, so I'd set status for requests that don't have this property. `api_spec.yaml` has 20+ endpoints

**Comment by Stranger6667** [comment]
> Does it mean that we will set status for all cases? 

Yes, you might need to do it conditionally only for cases that require this property

**Comment by KissPeter** [comment]
%00 causes issues, is there a way to instruct  schemathesis  or hypothesis  to exclude it?

**Comment by Stranger6667** [comment]
Hi @KissPeter, it is one of the thing I will work on for the next Schemathesis release. Currently there are only options like changing the schema or using a hook, but Hypothesis recently added support for excluding null bytes directly, so I am going to integrate it here too.

**Comment by Stranger6667** [comment]
See this comment for a solution that will work for the current version - https://github.com/schemathesis/schemathesis/issues/1668#issuecomment-1427021190
